pyfarm/master/user_interface/software.py

```diff
@@ -0,0 +1,191 @@
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+try:
+    from httplib import NOT_FOUND, SEE_OTHER, BAD_REQUEST
+except ImportError:  # pragma: no cover
+    from http.client import NOT_FOUND, SEE_OTHER, BAD_REQUEST
+
+from flask import render_template, request, redirect, url_for, flash
+
+from sqlalchemy import or_
+from sqlalchemy.sql.expression import func
+
+from pyfarm.models.software import (
+    Software, SoftwareVersion, JobSoftwareRequirement,
+    JobTypeSoftwareRequirement)
+from pyfarm.master.application import db
+
+def software():
+    software = Software.query.order_by("software desc").all()
+
+    return render_template("pyfarm/user_interface/software.html",
+                           software=software)
+
+def software_item(software_id):
+    software = Software.query.filter_by(id=software_id).first()
+    if not software:
+        return (render_template("pyfarm/error.html",
+                                error="Software %s not found" % software_id),
+                NOT_FOUND)
+
+    versions = SoftwareVersion.query.filter_by(
+        software=software).order_by("rank desc").all()
+
+    max_rank, = db.session.query(func.max(SoftwareVersion.rank)).\
+        filter(SoftwareVersion.software == software).first()
+
+    if not max_rank:
+        max_rank = 0
+
+    next_rank = int((round((float(max_rank) + 100) / 100) * 100))
+
+    return render_template("pyfarm/user_interface/software_item.html",
+                           software=software, versions=versions,
+                           next_rank=next_rank)
+
+def add_software():
+    software = Software(software=request.form["software"])
+
+    db.session.add(software)
+    db.session.commit()
+
+    flash("Software %s created" % software.software)
+
+    return redirect(url_for("single_software_ui", software_id=software.id),
+                    SEE_OTHER)
+
+def remove_software(software_id):
+    software = Software.query.filter_by(id=software_id).first()
+    if not software:
+        return (render_template("pyfarm/error.html",
+                                error="Software %s not found" % software_id),
+                NOT_FOUND)
+
+    job_requirements = JobSoftwareRequirement.query.filter(
+        JobSoftwareRequirement.software == software).all()
+    jobtype_requirements = JobTypeSoftwareRequirement.query.filter(
+        JobTypeSoftwareRequirement.software == software).all()
+    if job_requirements or jobtype_requirements:
+        error = ("This software cannot be deleted, it is still in use by the "
+                 "following jobs or jobtypes: ")
+        first = True
+        for job_requirement in job_requirements:
```

Comment:
You could probably simplify this logic by appending to a list instead of the original error string.  Then you'd just do `", ".join(errors)`